### ISSUE TITLE
#[allow(clippy::option_option)] not working when paired with #[serde(deserialize_with)]

### ISSUE THREAD

**Issue by schungx** [rust-lang/rust-clippy#4298]
Using: VS Code with RLS, Windows

This produces clippy warning on `option_option`:

        #[allow(clippy::option_option)]
        #[serde(deserialize_with = "func")]
        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(default)]
        #[serde(borrow)]
        foo: Option<Option<Cow<'a, str>>>,

~~~
consider using `Option<T>` instead of `Option<Option<T>>` or a custom enum if you need to distinguish all 3 cases.
~~~

However, removing `#[serde(deserialize_with ... )]` makes it work:

        #[allow(clippy::option_option)]
        #[serde(skip_serializing_if = "Option::is_none")]
        #[serde(default)]
        #[serde(borrow)]
        foo: Option<Option<Cow<'a, str>>>,

I suspect `#[serde(serialize_with ...)]` will also trip up this feature.  So something about `serde`'s `serialize_with` and `deserialize_with` attributes conflict with `option_option`.

**Comment by flip1995** [comment]
Minimal reproducer:
```rust
use serde::{Deserialize, Deserializer, Serialize}; // 1.0.94
use std::borrow::Cow;

#[derive(Serialize, Deserialize)]
struct Foo<'a> {
    #[allow(clippy::option_option)]
    #[serde(deserialize_with = "func")]
    #[serde(skip_serializing_if = "Option::is_none")]
    #[serde(default)]
    #[serde(borrow)]
    foo: Option<Option<Cow<'a, str>>>,
}

#[allow(clippy::option_option)]
fn func<'a, D>(_: D) -> Result<Option<Option<Cow<'a, str>>>, D::Error>
where
    D: Deserializer<'a>,
{
    Ok(Some(Some(Cow::Borrowed("hi"))))
}

fn main() {}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=db3c253dfb93476b696629e97a80b572)

The warning messages are because of the expansion of `deserialize_with`. This probably just needs a macro check.

**Comment by montrivo** [comment]
Seems to be fixed, @flip1995 reproduction example does not trigger any lint.

**Comment by schungx** [comment]
> Seems to be fixed

Yes, it looks like it fixed itself a while ago.

Closing this one.

**Comment by montrivo** [comment]
The `uitest` suite on this test case still triggers incorrectly. Maybe a serde version difference? This issue could stay open.
